File: levelledmobs-plugin-bukkit/src/main/java/io/github/arcaneplugins/levelledmobs/bukkit/logic/function/process/action/impl/RunFunctionAction.java
package io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.impl;

import io.github.arcaneplugins.levelledmobs.bukkit.LevelledMobs;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.LogicHandler;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.context.Context;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.LmFunction;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.Process;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.Action;
import io.github.arcaneplugins.levelledmobs.bukkit.util.Log;
import java.util.Objects;
import java.util.Optional;
import javax.annotation.Nonnull;
import org.jetbrains.annotations.NotNull;
import org.spongepowered.configurate.CommentedConfigurationNode;

public class RunFunctionAction extends Action {

    /* vars */

    private final String otherFuncId;
    private boolean sentError = false;

    /* constructors */

    public RunFunctionAction(
        @NotNull Process process,
        @NotNull CommentedConfigurationNode node
    ) {
        super(process, node);
        this.otherFuncId = Objects.requireNonNull(
            getActionNode().node("id").getString(),
            "RunFunctionAction did not specify valid ID of function to run"
        );
    }

    /* methods */

    @Override
    public void run(Context context) {
        final boolean potentiallyCircularFunction = context.getLinkedFunctions().stream()
            .anyMatch(otherFunction -> otherFunction.getIdentifier().equals(getOtherFuncId()));

        final boolean useCircularFunctionDependencyDetection = LevelledMobs.getInstance()
            .getConfigHandler().getSettingsCfg()
            .getRoot().node("advanced", "circular-function-dependency-detection")
            .getBoolean(false);

        if(useCircularFunctionDependencyDetection && potentiallyCircularFunction) {
            Log.sev(String.format(
                "Blocked potentially recursive call to run function '%s' in process '%s' (parent " +
                    "function '%s'). This protection can be disabled - be advised that recursive " +
                    "calls can result in memory leaks. LM will call 'exit-all' on the cause. " +
                    "This message will only appear once.",
                getOtherFuncId(),
                getParentProcess().getIdentifier(),
                getParentProcess().getParentFunction().getIdentifier()
            ), true);
            getParentProcess().getParentFunction().exitAll(context);
            return;
        }

        final Optional<LmFunction> functionToRunOpt = LogicHandler.getFunctions().stream()
            .filter(otherFunction -> otherFunction.getIdentifier().equals(getOtherFuncId()))
            .findFirst();

        if(functionToRunOpt.isEmpty()) {
            if(hasSentError())
                return;

            Log.sev(String.format(
                "Unable to run function '%s' from process '%s' in function '%s' as function '%s' " +
                    "does not exist.",
                getOtherFuncId(),
                getParentProcess().getIdentifier(),
                getParentProcess().getParentFunction().getIdentifier(),
                getOtherFuncId()
            ), true);

            setHasSentError(true);
        } else {
            if(!potentiallyCircularFunction)
                context.withLinkedFunction(getParentProcess().getParentFunction());

            functionToRunOpt.get().run(context, false);
        }
    }

    /* getters and setters */

    public boolean hasSentError() {
        return sentError;
    }

    public void setHasSentError(final boolean state) {
        this.sentError = state;
    }

    @Nonnull
    public String getOtherFuncId() {
        return otherFuncId;
    }
}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/logic/function/process/action/impl/RunFunctionAction.kt
package io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.impl

import io.github.arcaneplugins.levelledmobs.bukkit.LevelledMobs
import io.github.arcaneplugins.levelledmobs.bukkit.logic.LogicHandler
import io.github.arcaneplugins.levelledmobs.bukkit.logic.context.Context
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.LmFunction
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.Process
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.Action
import io.github.arcaneplugins.levelledmobs.bukkit.util.Log.sev
import org.spongepowered.configurate.CommentedConfigurationNode
import java.util.*

class RunFunctionAction(
    val process: Process,
    val node: CommentedConfigurationNode
): Action(process, node) {
    val otherFuncId: String? = actionNode.node("id").string
    var sentError = false

    init {
        require(
            otherFuncId != null
        ) { "RunFunctionAction did not specify valid ID of function to run" };
    }

    override fun run(context: Context) {

        val potentiallyCircularFunction = context.linkedFunctions.stream()
            .anyMatch { otherFunction: LmFunction -> otherFunction.identifier == otherFuncId }

        val useCircularFunctionDependencyDetection: Boolean = LevelledMobs.lmInstance
            .configHandler.settingsCfg
            .root!!.node("advanced", "circular-function-dependency-detection")
            .getBoolean(false)

        if (useCircularFunctionDependencyDetection && potentiallyCircularFunction) {
            sev(
                "Blocked potentially recursive call to run function '$otherFuncId' in process '${parentProcess.identifier}' (parent " +
                        "function '${parentProcess.parentFunction.identifier}'). This protection can be disabled - be advised that recursive " +
                        "calls can result in memory leaks. LM will call 'exit-all' on the cause. " +
                        "This message will only appear once.",
    true
            )
            parentProcess.parentFunction.exitAll(context)
            return
        }

        val functionToRunOpt: Optional<LmFunction> = LogicHandler.lmFunctions.stream()
            .filter { otherFunction -> otherFunction.identifier == otherFuncId }
            .findFirst()

        if (functionToRunOpt.isEmpty) {
            if (sentError) return
            sev(
                "Unable to run function '$otherFuncId' from process '${parentProcess.identifier}' in function '${parentProcess.parentFunction.identifier}' as function '$otherFuncId' " +
                        "does not exist.",
                true
            )
            sentError = true
        } else {
            if (!potentiallyCircularFunction) context.withLinkedFunction(parentProcess.parentFunction)
            functionToRunOpt.get().run(context, false)
        }
    }
}